File: src/dolphin_context_actions/ui.py
import os
import shutil
import subprocess
import sys

QDBUS = next(
    (n for n in ("qdbus-qt5", "qdbus", "qdbus6") if shutil.which(n)),
    None,
)


def kdialog(*args) -> subprocess.CompletedProcess:
    try:
        r = subprocess.run(
            ["kdialog"] + list(args),
            capture_output=True,
            text=True,
            env=os.environ.copy(),
        )
    except OSError as exc:
        # Absent on headless systems; callers inspect returncode/stdout, so
        # return a failed result instead of crashing.
        print(f"kdialog unavailable: {exc}. Args: {args}", file=sys.stderr)
        return subprocess.CompletedProcess(["kdialog", *args], 1, "", str(exc))
    if r.returncode != 0 and not shutil.which("kdialog"):
        print(f"kdialog not found. Args: {args}", file=sys.stderr)
    return r


def notify(title: str, msg: str, icon: str = "document-convert"):
    # Fire and forget. notify-send blocks indefinitely when no notification
    # daemon is running, which used to stall the caller mid-operation.
    try:
        subprocess.Popen(
            ["notify-send", "-i", icon, "-a", "Context Actions", title, msg],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
    except OSError:
        pass  # notify-send not installed; notifications are advisory


def error_dialog(title: str, msg: str):
    kdialog("--title", title, "--error", msg)


def info_dialog(title: str, msg: str, width: int = 0, height: int = 0):
    args = ["--title", title, "--msgbox", msg]
    if width > 0 and height > 0:
        args.extend(["--geometry", f"{width}x{height}"])
    kdialog(*args)


def confirm_dialog(title: str, msg: str) -> bool:
    r = kdialog("--title", title, "--yesno", msg)
    return r.returncode == 0


def menu_dialog(title: str, prompt: str, items: list[tuple[str, str]]) -> str | None:
    args = ["--title", title, "--menu", prompt]
    for key, label in items:
        if not label:
            continue  # kdialog has no separators; a blank row is just a dead entry
        args += [key, label]
    r = kdialog(*args)
    if r.returncode != 0:
        return None
    return r.stdout.strip()


def input_dialog(title: str, prompt: str, text: str = "") -> str | None:
    r = kdialog("--title", title, "--inputbox", prompt, text)
    if r.returncode != 0:
        return None
    return r.stdout.strip()


def pbar_open(title: str, label: str) -> tuple | None:
    r = kdialog("--title", title, "--progressbar", label, "100")
    parts = r.stdout.strip().split()
    if len(parts) >= 2:
        return (parts[0], parts[1])
    if len(parts) == 1:
        return (parts[0], "/")
    return None


def pbar_set(handle: tuple | None, value: int, label: str | None = None) -> bool:
    if not handle or not QDBUS:
        return True
    service, path = handle
    # This runs every ~0.5s from the ffmpeg progress-polling loop, so a qdbus
    # call that hangs (unresponsive D-Bus daemon) would otherwise stall that
    # loop indefinitely. Treat a timeout as "can't reach the progress UI right
    # now", not "user clicked cancel" -- return True (keep converting) rather
    # than aborting a healthy conversion over an unrelated D-Bus hiccup.
    try:
        r = subprocess.run(
            [QDBUS, service, path, "Set", "", "value", str(value)],
            capture_output=True,
            timeout=5,
        )
    except subprocess.TimeoutExpired:
        return True
    if r.returncode != 0:
        return False
    if label is not None:
        try:
            subprocess.run(
                [QDBUS, service, path, "setLabelText", label],
                capture_output=True,
                timeout=5,
            )
        except subprocess.TimeoutExpired:
            pass
    return True


def pbar_close(handle: tuple | None):
    if handle and QDBUS:
        try:
            subprocess.run([QDBUS, handle[0], handle[1], "close"], capture_output=True, timeout=5)
        except subprocess.TimeoutExpired:
            pass

File: src/ui.rs
use std::env;
use std::io;
use std::os::unix::process::ExitStatusExt;
use std::process::{Command, Output, Stdio};

static QDBUS_CANDIDATES: &[&str] = &["qdbus-qt5", "qdbus", "qdbus6"];

pub fn gui_blocked() -> bool {
    env::var_os("DOLPHIN_CONTEXT_ACTIONS_HEADLESS").is_some()
        || env::var_os("PYTEST_CURRENT_TEST").is_some()
}

fn completed(_args: &[&str], status: i32, stdout: &str, stderr: &str) -> Output {
    Output {
        status: ExitStatusExt::from_raw(status << 8),
        stdout: stdout.as_bytes().to_vec(),
        stderr: stderr.as_bytes().to_vec(),
    }
}

pub fn kdialog(args: &[&str]) -> Output {
    if gui_blocked() {
        eprintln!("kdialog skipped (headless): {args:?}");
        return completed(args, 1, "", "headless");
    }
    match Command::new("kdialog").args(args).output() {
        Ok(output) => {
            if !output.status.success() && which::which("kdialog").is_err() {
                eprintln!("kdialog not found. Args: {args:?}");
            }
            output
        }
        Err(exc) => {
            eprintln!("kdialog unavailable: {exc}. Args: {args:?}");
            completed(args, 1, "", &exc.to_string())
        }
    }
}

pub fn notify(title: &str, msg: &str, icon: &str) {
    if gui_blocked() {
        return;
    }
    let _ = Command::new("notify-send")
        .args(["-i", icon, "-a", "Context Actions", title, msg])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn();
}

pub fn error_dialog(title: &str, msg: &str) {
    let _ = kdialog(&["--title", title, "--error", msg]);
}

pub fn info_dialog(title: &str, msg: &str, width: i32, height: i32) {
    let mut args = vec!["--title", title, "--msgbox", msg];
    let geometry = format!("{width}x{height}");
    if width > 0 && height > 0 {
        args.extend(["--geometry", &geometry]);
    }
    let _ = kdialog(&args);
}

pub fn confirm_dialog(title: &str, msg: &str) -> bool {
    kdialog(&["--title", title, "--yesno", msg]).status.success()
}

pub fn menu_dialog(title: &str, prompt: &str, items: &[(&str, &str)]) -> Option<String> {
    let mut args = vec!["--title".to_string(), title.to_string(), "--menu".to_string(), prompt.to_string()];
    for (key, label) in items {
        if label.is_empty() {
            continue;
        }
        args.push((*key).to_string());
        args.push((*label).to_string());
    }
    let borrowed: Vec<&str> = args.iter().map(String::as_str).collect();
    let result = kdialog(&borrowed);
    if !result.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&result.stdout).trim().to_string())
}

pub fn input_dialog(title: &str, prompt: &str, text: &str) -> Option<String> {
    let result = kdialog(&["--title", title, "--inputbox", prompt, text]);
    if !result.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&result.stdout).trim().to_string())
}

pub fn pbar_open(title: &str, label: &str) -> Option<(String, String)> {
    let result = kdialog(&["--title", title, "--progressbar", label, "100"]);
    let stdout = String::from_utf8_lossy(&result.stdout);
    let parts: Vec<&str> = stdout.split_whitespace().collect();
    if parts.len() >= 2 {
        Some((parts[0].to_string(), parts[1].to_string()))
    } else if parts.len() == 1 {
        Some((parts[0].to_string(), "/".to_string()))
    } else {
        None
    }
}

fn qdbus() -> Option<String> {
    QDBUS_CANDIDATES
        .iter()
        .find_map(|name| which::which(name).ok().map(|p| p.display().to_string()))
}

pub fn pbar_set(handle: Option<&(String, String)>, value: i32, label: Option<&str>) -> bool {
    let Some((service, path)) = handle else {
        return true;
    };
    let Some(qdbus) = qdbus() else {
        return true;
    };
    let value_s = value.to_string();
    let mut cmd = Command::new(&qdbus);
    cmd.args([service, path, "Set", "", "value", &value_s]);
    match cmd.output() {
        Ok(out) if out.status.success() => {}
        Ok(_) => return false,
        Err(err) if err.kind() == io::ErrorKind::TimedOut => return true,
        Err(_) => return false,
    }
    if let Some(label) = label {
        let _ = Command::new(&qdbus)
            .args([service.as_str(), path.as_str(), "setLabelText", label])
            .output();
    }
    true
}

pub fn pbar_close(handle: Option<&(String, String)>) {
    let Some((service, path)) = handle else {
        return;
    };
    let Some(qdbus) = qdbus() else {
        return;
    };
    let _ = Command::new(qdbus).args([service, path, "close"]).output();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kdialog_does_not_exec_when_headless() {
        env::set_var("DOLPHIN_CONTEXT_ACTIONS_HEADLESS", "1");
        let result = kdialog(&["--title", "Dolphin Context Actions", "--error", "nope"]);
        env::remove_var("DOLPHIN_CONTEXT_ACTIONS_HEADLESS");
        assert!(!result.status.success());
        assert_eq!(String::from_utf8_lossy(&result.stderr), "headless");
    }

    #[test]
    fn gui_blocked_explicit_flag() {
        env::remove_var("PYTEST_CURRENT_TEST");
        env::remove_var("DOLPHIN_CONTEXT_ACTIONS_HEADLESS");
        assert!(!gui_blocked());
        env::set_var("DOLPHIN_CONTEXT_ACTIONS_HEADLESS", "1");
        assert!(gui_blocked());
        env::remove_var("DOLPHIN_CONTEXT_ACTIONS_HEADLESS");
    }
}
